### PR TITLE
Ghcjs support

### PR DESCRIPTION
**Pull request by benkolera**
This addresses the request in #54 for mmark to work in ghcjs. 

It just always returns an empty object on ghcjs, but we could change it to an error instead if that's better.

I really hate the CPP that litters the tests now (because of the different hashmap impls in ghcjs and ghc reversing the attribute ordering), but the tests do pass as they are in both ghcjs and ghc. What may be better:

- Instead of basing the tests on strings, set the expectations with Lucid.Html so that we can guarantee the ordering will match up
- Patch Lucid to sort the columns before writing them to the builder. Or use Data.Map instead of HashMap. 
- Or leave things how they are if the tests are pretty stable.

I can remove the shell.nix if you'd like, but it makes it really easy for a nix user to switch between ghc and ghcjs testing.

### CODE REVIEW

**Comment by mrkkrp** [comment]
I remember about this one, will review once I have time.

**Comment by benkolera** [comment]
No rush @mrkkrp . The application that I'm using this in is in a nix project, so it doesn't anything to point to my fork until you have time.

File: tests/Text/MMarkSpec.hs [hunk 1]
```diff
@@ -382,10 +383,18 @@ spec = parallel $ do
     context "4.7 Link reference definitions" $ do
       it "CM159" $
         "[foo]: /url \"title\"\n\n[foo]" ==->
+#ifdef ghcjs_HOST_OS
+          "<p><a title=\"title\" href=\"/url\">foo</a></p>\n"
```
Comment (mrkkrp) [inline tests/Text/MMarkSpec.hs]: This seems to be unrelated to the actual change in this PR. Do you have any idea why the order of attributes is different with GHCJS?

File: shell.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# You can get a shell with nix-shell --arg ghcjs <true|false>
```
Comment (mrkkrp) [inline shell.nix]: I prefer not to keep files which tend to change a lot as time passes, such as `stack.yaml` and `shell.nix` in repos of libraries (as opposed to repos of actual projects, where it is fine). So I'd rather drop this one.

File: tests/Text/MMarkSpec.hs (hunk 1 above)
Comment (benkolera) [reply tests/Text/MMarkSpec.hs]: Lucid is using a hashmap, which is unordered by design. I presume that JS is using a different hashing function. Having lucid use an ordered map instead of a unordered one would resolve the difference.

File: shell.nix (hunk 2 above)
Comment (benkolera) [reply shell.nix]: Can delete this, no problems.

**Comment by mrkkrp** [comment]
Ideally if we adding support for GHCJS, I'd like to add a CI job that checks that this works, otherwise it doesn't really make sense and will eventually break.

**Comment by benkolera** [comment]
There seems to be a decent path for getting ghcjs in Travis. I'll give that a whirl. https://github.com/nomeata/ghcjs2gh-pages/blob/master/.travis.yml

**Comment by benkolera** [comment]
OK, @mrkkrp ! I think I have something that is closer to acceptable. 

See the new `(##->)` that can be used for testing with a Lucid.Html as the expectation rather than a string. I only applied this to the tests that actually fail due to the different ghcjs hashing to keep your tests closer to test suite that you borrowed from. An example: https://github.com/mmark-md/mmark/pull/61/files#diff-eee44bc1c34d7345749f9227878bcf9bR389 

Given that you only have one spec anyway, it was easier to get rid on hspec-discover than making it work with cabal --ghcjs. I can make it work if you want, but it requires a hack in the .travis.yaml because cabal --ghcjs install doesn't do the right thing. :slightly_frowning_face: 

But travis is building the GHCJs and running the tests for ghcjs 8.4, it is passing and the tests are much tidier now, so I think that we're pretty close now. Let me know what you think. The only thing that I don't like is that it links the tests twice, which for ghcjs is really slow (adds a waste of ~110seconds to the build).

**Comment by mrkkrp** [comment]
Sorry for the delay, I'm just still not sure if GHCJS support is worth the trouble.

**Comment by benkolera** [comment]
What's the trouble at this point though?

**Comment by srid** [comment]
It would be nice to be able to parse and render markdown in the browser. Perhaps all those CPP can be moved to a separate module (which would export the wrapped yaml functions)?

**Comment by srid** [comment]
I can understand the unnecessary burden of taking up GHCJS support; but what if you didn't have to (thereby letting the community handle it)? :-)

**Comment by mrkkrp** [comment]
I'll merge this in once I have time, but I won't be available to solve GHCJS-specific issues, so this will be on you guys.

**Comment by benkolera** [comment]
Awesome. Thank you! Yeah very much send me any ghcjs issues my way. Part of
my work at the Queensland (Australia) FP Lab (QFPL) is dedicated to FP OSS
work, so I'm more than happy to deal with GHCjs issues that arise.

If you go to make changes later and the ghcjs gets in the way, let me know
of that too. We may be able to tweak things so that they are less annoying.

File: .travis.yml [hunk 3]
```diff
@@ -4,33 +4,49 @@ sudo: false
 
 cache:
   directories:
-    - ~/.cabal/packages
-    - ~/.cabal/store
+    - $HOME/.ghc/
```
Comment (mrkkrp) [inline .travis.yml]: Why do you exclude `~/.cabal/packages` and why do you prefer `$HOME` to `~/`?

I think `~/.cabal/packages` used to do the trick. I talked to HVR and he said that only `packages` and `store` are necessary. What's the reason for inclusion each of these directories?

Comment (benkolera) [reply .travis.yml]: If HVR says that should be OK, then I'll revert it. I only went that way because that's how:

https://github.com/nomeata/ghcjs2gh-pages/blob/master/.travis.yml#L22

did it. $HOME vs ~ is meaningless, but the directory changes sound like that may not be ideal.

Comment (benkolera) [reply .travis.yml]: Fixed.

**Comment by mrkkrp** [comment]
I was on a vacation trip the whole last week. I'll get back to this now.

**Comment by mrkkrp** [comment]
This is merged. I'll cut 0.0.7.0 soon.

**Comment by benkolera** [comment]
That's OK. Thanks @mrkkrp ! :slightly_smiling_face:

**Comment by mrkkrp** [comment]
Version 0.0.7.0 is on Hackage.

**Comment by augyg** [comment]
I'm getting very strange behaviour specifically when I work with template haskell in GHCJS 

the same input string will work completely fine with ghc and then fail like so in Template Haskell GHCJS

image


which appears to be from this section:

image

**Comment by augyg** [comment]
image

**Comment by augyg** [comment]
Also failing even when not using template haskell but just generically with ghcjs
